src/parthenon_arrays.hpp

```diff
@@ -0,0 +1,309 @@
+#include <type_traits>
+#include <utility> // make_pair
+#include <vector>
+
+// Kokkos Headers
+#include <Kokkos_Core.hpp>
+
+// Parthenon++ headers
+#include "kokkos_abstraction.hpp"
+
+// Macro for automatically creating a useful name
+#define PARARRAY_TEMP\
+  "ParArrayNDGeneric:"+std::string(__FILE__)+":"+std::to_string(__LINE__)
+
+namespace parthenon {
+
+#define SLC0 std::make_pair(0,1)
+
+// API designed with Container = Kokkos::View<T******> in mind
+template <typename Container>
+class ParArrayNDGeneric {
+ public:
+  using index_pair_t = std::pair<size_t,size_t>;
+
+  KOKKOS_FUNCTION
+  ParArrayNDGeneric() = default;
+  explicit ParArrayNDGeneric(const std::string& label,
+                      int nx6, int nx5, int nx4, int nx3, int nx2, int nx1)
+    : d6d_(label,nx6,nx5,nx4,nx3,nx2,nx1)
+  { }
+  ParArrayNDGeneric(const std::string& label,
+             int nx5, int nx4, int nx3, int nx2, int nx1)
+    : d6d_(label,1,nx5,nx4,nx3,nx2,nx1)
+  { }
+  ParArrayNDGeneric(const std::string& label,
+             int nx4, int nx3, int nx2, int nx1)
+    : d6d_(label,1,1,nx4,nx3,nx2,nx1)
+  { }
+  ParArrayNDGeneric(const std::string& label, int nx3, int nx2, int nx1)
+    : d6d_(label,1,1,1,nx3,nx2,nx1)
+  { }
+  ParArrayNDGeneric(const std::string& label, int nx2, int nx1)
+    : d6d_(label,1,1,1,1,nx2,nx1)
+  { }
+  ParArrayNDGeneric(const std::string& label, int nx1)
+    : d6d_(label,1,1,1,1,1,nx1)
+  { }
+  explicit ParArrayNDGeneric(const Container& v)
+    : d6d_(v)
+  {}
+
+  // legacy functions, as requested for backwards compatibility
+  // with athena++ patterns
+  void NewParArrayND(int nx1) {
+    d6d_ = Container("ParArray1D",1,1,1,1,1,nx1);
+  }
+  void NewParArrayND(int nx2, int nx1) {
+    d6d_ = Container("ParArray2D",1,1,1,1,nx2,nx1);
+  }
+  void NewParArrayND(int nx3, int nx2, int nx1) {
+    d6d_ = Container("ParArray3D",1,1,1,nx3,nx2,nx1);
+  }
+  void NewParArrayND(int nx4, int nx3, int nx2, int nx1) {
+    d6d_ = Container("ParArray4D",1,1,nx4,nx3,nx2,nx1);
+  }
+  void NewParArrayND(int nx5, int nx4, int nx3, int nx2, int nx1) {
+    d6d_ = Container("ParArray5D",1,nx5,nx4,nx3,nx2,nx1);
+  }
+  void NewParArrayND(int nx6, int nx5, int nx4, int nx3, int nx2, int nx1) {
+    d6d_ = Container("ParArray6D",nx6,nx5,nx4,nx3,nx2,nx1);
+  }
+
+  KOKKOS_INLINE_FUNCTION __attribute__((nothrow))
+  ParArrayNDGeneric(const ParArrayNDGeneric<Container>& t) = default;
+  KOKKOS_INLINE_FUNCTION __attribute__((nothrow))
+  ~ParArrayNDGeneric() = default;
+  KOKKOS_INLINE_FUNCTION __attribute__((nothrow))
+  ParArrayNDGeneric<Container>
+  &operator= (const ParArrayNDGeneric<Container> &t) = default;
+  KOKKOS_INLINE_FUNCTION __attribute__((nothrow))
+  ParArrayNDGeneric(ParArrayNDGeneric<Container>&& t) = default;
+  KOKKOS_INLINE_FUNCTION __attribute__((nothrow))
+  ParArrayNDGeneric<Container> &operator= (ParArrayNDGeneric<Container> &&t) = default;
+
+  // function to get the label
+  inline const std::string label() const {
```

Comment:
`inline` in this context is redundant - any function defined in-line (heh) in the class definition is default `inline`.

Reply:
Good catch. Thanks.

Reply:
Ah, actually I just checked. `Kokkos::View.label()` returns a new object, not a reference. So returning a reference to it doesn't work. I'm going to change it to just 
```C++
std::string label() const
```

Reply:
OK - that works, too!